adds splitBy extension method on scala.collection.Iterator

jeantil opened the pull request:
`Iterator#splitBy` constructs an iterator where consecutive
elements of the original iterator are accumulated as long as the output
of a key function for each element doesn't change. They are emitted as
an Iterable as soon as the key function changes. 

This operation makes sense as soon as you are trying to process an
iterator where you know the elements will be sorted in a certain way and
you need to group them without loading all the data in memory.

For instance
 * processing a file where the ordering is guaranteed but the file
 doesn't fit in the heap,
 * processing a streaming resultset where the underlying database
 guarantees the ordering because of a sort clause.

The same operation is added on Iterables with the difference that the
specific container type of the input is preserved for both collection 
levels of the output, thus 
* `Set(1,2,3).splitBy(identity)` returns 
`Set(Set(1), Set(2), Set(3))` 
* `Vector(1,2,3).splitBy(identity)` returns 
`Vector(Vector1), Vector2), Vector3))` 
* etc.

thebignet commented on src/main/scala/scala/collection/decorators/IteratorDecorator.scala:
```diff
@@ -72,4 +72,47 @@ class IteratorDecorator[A](val `this`: Iterator[A]) extends AnyVal {
+    */
+  def groupUntilChanged[K](f: A => K): Iterator[Iterable[A]] =
+    new AbstractIterator[Seq[A]] {
+      private var hd: Option[A] = readHead()
```
This variable could be named `maybeHead` or something that shows a better intent

jeantil replied:
Variables with the same kind of usage are named like this in the std lib (search var hd in https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/Iterator.scala) so I chose to remain consistent. Arguably I should have stayed consistent all the way through and used a nullable variable and a boolean but I didn't feel very confortable with that)

thebignet commented on src/main/scala/scala/collection/decorators/IteratorDecorator.scala:
```diff
@@ -72,4 +72,47 @@ class IteratorDecorator[A](val `this`: Iterator[A]) extends AnyVal {
+    */
+  def groupUntilChanged[K](f: A => K): Iterator[Iterable[A]] =
+    new AbstractIterator[Seq[A]] {
+      private var hd: Option[A] = readHead()
```
Is it necessary to have such a variable ? It's used in `next()` and `hasNext`. For `next()`, the match could be directly on `readHead()` whereas `hasNext` could rely on `this.hasNext` instead. This would delay the `next` action to when `next` is actually called, at least for the first call on `next`

jeantil replied:
I don't see a way around the variable. We need to peek ahead in the underlying iterator in order to know if we should continue accumulating or if we should start another group. Having read one ahead, it has to be stored locally or it would be lost.

thebignet replied:
True, I did not see that point. The `seq.toVector` can only be done if we know that the next element changes key. Too bad we could not just rely on the state of the underlying Iterable to do that.

jeantil commented:
I made cosmetic changes after the first review : 
* inlined local val `key` as it was used only once
* replace custom-defined `readHead` with `Iterator#nextOption` since the behaviour is exactly the same.

jeantil commented:
Hello 

@julienrf I don't mind if this isn't a priority, but just want to make sure this PR was submitted at the right place (not to mention I would love feedback on the actual implementation :) )

cheers

julienrf commented:
Hello Jean! Sorry for being silent. Yes, the PR looks great! We are still unsure about what will happen to `scala-collection-contrib`, though. I will come back to you soon. Is that ok for you?

jeantil commented:
Hi Julien, thats perfectly ok for me. I'll keep monitoring the thread here and at contributors.scala-lang.org

cheers :)

joshlemer commented on src/main/scala/scala/collection/decorators/IteratorDecorator.scala:
```diff
@@ -72,4 +72,44 @@ class IteratorDecorator[A](val `this`: Iterator[A]) extends AnyVal {
+              seq = seq ++ hd
+              hd = `this`.nextOption()
+            }
+            seq.toVector
```
`seq = seq ++ hd` will create a clone of `mutable.Buffer(head)` and append the element to it. And then at the end `toVector` will have to copy again. It would be much more efficient to just create a `Vector.newBuilder[A]` and append each element to that.

jeantil replied:
ack

joshlemer commented on src/main/scala/scala/collection/decorators/IteratorDecorator.scala:
```diff
@@ -72,4 +72,44 @@ class IteratorDecorator[A](val `this`: Iterator[A]) extends AnyVal {
+        hd match {
+          case None => Iterator.empty.next()
+          case Some(head) =>
+            hd = `this`.nextOption()
```
Most traversals in the standard library use the `hasNext`/`next` calls to avoid allocating `Option` objects, we could do the same thing here :-). Then we can also avoid the function allocation inside the `exists` call on line `107` 😄

jeantil replied:
I understand however that would force me to use null for hd when hasNext is `false` is that acceptable ?

joshlemer replied:
No, this won't work because `null` can be a valid element of `A`. I'm also just realizing now that evaluating ``` `this`.nextOption() ``` at all is probably not kosher though, since it violates the laziness of `Iterator`s. I reckon the first call to ``` `this`.next() ```shouldn't happen until the first call tho `next()`.

jeantil replied:
when I talk about using `null`, I am thinking of somethin akin to   https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/Iterator.scala#L122

joshlemer replied:
Ah yes, that'll work!

jeantil replied:
I have tried to address your point, I will push a fixup commit with my attempt at this... I have tried to keep it as simple as possible, it passes the tests but I find it much much less satisfying than the original implementation.

joshlemer commented on src/main/scala/scala/collection/decorators/IteratorDecorator.scala:
```diff
@@ -72,4 +72,44 @@ class IteratorDecorator[A](val `this`: Iterator[A]) extends AnyVal {
+    * @return an iterator of sequences of the consecutive elements with the
+    *         same key in the original iterator
+    */
+  def groupUntilChanged[K](f: A => K): Iterator[Iterable[A]] =
```
Sorry for the nitpick / bikeshedding but maybe something shorter like `chunkBy`?

joshlemer replied:
Other possibilities:
`segmentBy`
`splitBy`
`sliceBy`

joshlemer replied:
In fact, since this is kind of a generalization of https://github.com/scala/scala-collection-contrib/pull/6, it probably makes most sense to call it `splitBy`

jeantil replied:
`splitBy` seems to be what [Mathematica calls it](https://reference.wolfram.com/language/ref/SplitBy.html)  sounds good to me.

jeantil replied:
I will push a fixup commit addressing the rename shortly

joshlemer commented:
If we're gonna add this `splitBy` method on Iterators, it probably makes sense to add it on the Iterables as well, which delegates to basically ``` `this`.iterator.splitBy(f).map(`this`.fromSpecific)``` or something like that. See for instance, how this PR adds both implementations https://github.com/scala/scala-collection-contrib/pull/18/files#diff-ea5808f2cdf7aad91c2879f38e526320R34

jeantil commented:
@joshlemer I gave a try to adding this to `IterableDecorator`.
I am not entirely sure what the resulting signature should be. I went for `Iterator[C[A]]` according to your hint so that `Vector(1,1,2,2,3,3).splitBy(identity)` would  return `Iterator[Vector[A]]` but I don't know if rewrapping the inner collections in a Vector actually brings much value.

julienrf commented:
@jeantil I think it should return at least `Iterable[CC[A]]`, but maybe `CC[CC[A]]` is a more obvious default?

ywg-jean commented:
Here is what I came up with for a quick shake up : 
```
def splitBy[K, That,CC](f: seq.A => K)(implicit bf: BuildFrom[C, seq.A, That], bff: BuildFrom[C,That,CC]):CC =
    bff.fromSpecific(coll)(seq(coll).iterator.splitBy(f).map(bf.fromSpecific(coll)))
```
this compiles, but then I get compilation errors on the tests because of ambiguous implicits 
```
[error] Note that implicit conversions are not applicable because they are ambiguous:
[error]  both method IterableDecorator in package decorators of type [C](coll: C)(implicit it: scala.collection.generic.IsIterable[C])scala.collection.decorators.IterableDecorator[C,it.type]
[error]  and method SeqDecorator in package decorators of type [C](coll: C)(implicit seq: scala.collection.generic.IsSeq[C])scala.collection.decorators.SeqDecorator[C,seq.type]
[error]  are possible conversion functions from scala.collection.immutable.Vector[A] to ?{def splitBy: ?}
[error]     val groupedVector:Vector[Vector[Nothing]] = Vector.empty.splitBy(identity)
[error]                                                        ^
```

joshlemer commented:
@julienrf 
for Iterators:

We could aim for consistency with the most closely related method on iterator, `grouped`, it would then return an `Iterator[immutable.Seq[A]] `

for Seq's:

Since it's not changing the underlying element type, I think it could return `CC[C]`:
`def splitBy[B](f: A => B): CC[C]`

This would also be more consistent with `def splitAt(n: Int): (C, C)`

jeantil commented:
Here is what I came up with for a quick shake up : 
```scala
def splitBy[K, That,CC](f: seq.A => K)(implicit bf: BuildFrom[C, seq.A, That], bff: BuildFrom[C,That,CC]):CC =
    bff.fromSpecific(coll)(seq(coll).iterator.splitBy(f).map(bf.fromSpecific(coll)))
```
this compiles, but then I get compilation errors on the tests because of ambiguous implicits 
```scala
[error] Note that implicit conversions are not applicable because they are ambiguous:
[error]  both method IterableDecorator in package decorators of type [C](coll: C)(implicit it: scala.collection.generic.IsIterable[C])scala.collection.decorators.IterableDecorator[C,it.type]
[error]  and method SeqDecorator in package decorators of type [C](coll: C)(implicit seq: scala.collection.generic.IsSeq[C])scala.collection.decorators.SeqDecorator[C,seq.type]
[error]  are possible conversion functions from scala.collection.immutable.Vector[A] to ?{def splitBy: ?}
[error]     val groupedVector:Vector[Vector[Nothing]] = Vector.empty.splitBy(identity)
[error]                                                        ^
```

But that won't work with CC[C] (as per  @joshlemer's proposal) 

I tried getting it to work on `Iterable` instead but I get in other kind of errors and I'm short on time to dig deeper :(

julienrf commented:
@jeantil is `splitBy` defined on both `IterableDecorator` and `SeqDecorator`? It should be defined only on one of those.

julienrf commented:
@joshlemer You raised good points ;)

> We could aim for consistency with the most closely related method on iterator, `grouped`, it would then return an `Iterator[immutable.Seq[A]] `

I don’t know what is the reason for `grouped` to have this return type, though.

> for Seq's:
> 
> Since it's not changing the underlying element type, I think it could return `CC[C]`

:+1:

ywg-jean commented:
I realize I may not have pushed everything, I had started adding it on IterableDecorator and it seems i never pushed the commit

jeantil commented:
I realize I may not have pushed everything, I had started adding it on IterableDecorator and it seems i never pushed the commit 
(sorry about the account juggling)

jeantil commented:
Hi,
I have squashed the iterations on `IteratoDecorator#splitBy` and changed `IterableDecorator#splitBy` to return `CC[CC[A]]` when it is given a `C[A]`
I switched from delegating to iterator to a specific implementation which 
I found easier to work with to achieve the expected type signature.

The corresponding tests have been updated too.

julienrf commented on src/main/scala/scala/collection/decorators/IterableDecorator.scala:
```diff
@@ -33,4 +33,46 @@ class IterableDecorator[C, I <: IsIterable[C]](coll: C)(implicit val it: I) {
 
+
+  /**
+    * Constructs an iterator where consecutive elements are accumulated as
```
We should replace “iterator” with “collection”

jeantil replied:
done :)

julienrf commented on src/main/scala/scala/collection/decorators/IterableDecorator.scala:
```diff
@@ -33,4 +33,46 @@ class IterableDecorator[C, I <: IsIterable[C]](coll: C)(implicit val it: I) {
+    * </pre>
+    * produces
+    * <pre>
+    * Iterator(Vector(1),
```
It should be `Vector(Vector(1),` instead of `Iterator(Vector(1),`

jeantil replied:
done

julienrf commented on src/main/scala/scala/collection/decorators/IterableDecorator.scala:
```diff
@@ -33,4 +33,46 @@ class IterableDecorator[C, I <: IsIterable[C]](coll: C)(implicit val it: I) {
+    * @return an iterator of sequences of the consecutive elements with the
+    *         same key in the original iterator
+    */
+  def splitBy[K, That, CC](f: it.A => K)(implicit bf: BuildFrom[C, it.A, That], bff: BuildFrom[C, That, CC]): CC = {
```
What about naming the type parameters `That1`, `That2`? Or maybe `C1` and `C2`?

jeantil replied:
done

julienrf commented on src/main/scala/scala/collection/decorators/IterableDecorator.scala:
```diff
@@ -33,4 +33,46 @@ class IterableDecorator[C, I <: IsIterable[C]](coll: C)(implicit val it: I) {
+    * @param f the function to compute a key for an element
+    * @tparam K the type of the computed key
+    * @return an iterator of sequences of the consecutive elements with the
+    *         same key in the original iterator
```
What about the following description?

> \@return a sequence of sequences of consecutive elements having the same key

jeantil replied:
changed for collection of collections

julienrf commented on src/main/scala/scala/collection/decorators/IterableDecorator.scala:
```diff
@@ -33,4 +33,46 @@ class IterableDecorator[C, I <: IsIterable[C]](coll: C)(implicit val it: I) {
+      init += ref
+      while (iterator.hasNext) {
+        val el = iterator.next();
+        if (f(el) != f(ref)) {
```
Should we put `f(ref)` in a val to compute it only once?

jeantil replied:
I reverted to delegating to iterator's implementation and the tests still pass. I guess I got confused with my attempts at having splitBy defined for both IterableDecorator and SeqDecorator ...

jeantil replied:
this becomes irrelevant with the revert to delegating to iterator.

julienrf commented on src/main/scala/scala/collection/decorators/IteratorDecorator.scala:
```diff
@@ -72,4 +72,63 @@ class IteratorDecorator[A](val `this`: Iterator[A]) extends AnyVal {
+          while (`this`.hasNext && hadSameKey) {
+            val el = `this`.next()
+            hdDefined = true
+            if (f(el) == f(hd)) {
```
Same comment here about `f(hd)`

jeantil replied:
That wouldn't work, `hd` needs to be mutable to honor the iterator contract. It is mutated on line 123. 

The best I could come up with is adding another mutable variable `hdKey` which is only updated when the key actually changes. 
I have added a commit to demonstrate it. I feel it does damage the readability but I understand it can potentially have a huge performance impact.

jeantil commented:
Hi Julien,
It's alright, I'm doing this on the side and I really appreciate the time
you spend for reviews.
I may not be able to address your latest comments before next week.

Le lun. 19 août 2019 à 10:32, Julien Richard-Foy <notifications@github.com>
a écrit :

> *@julienrf* commented on this pull request.
>
> Thanks *a lot* for your patience, Jean!
>
> Doing the latter, I switched from delegating to iterator to a specific
> implementation which
> I found easier to work with to achieve the expected type signature.
>
> I understand, however the drawback is that your implementation is strict
> by default. I think it would be nice if it was lazy on LazyList.
> ------------------------------
>
> In src/main/scala/scala/collection/decorators/IterableDecorator.scala
> <https://github.com/scala/scala-collection-contrib/pull/4#discussion_r315086758>
> :
>
> > @@ -33,4 +33,46 @@ class IterableDecorator[C, I <: IsIterable[C]](coll: C)(implicit val it: I) {
>
>    def lazyFoldRight[B](z: B)(op: it.A => Either[B, B => B]): B =
>
>      it(coll).iterator.lazyFoldRight(z)(op)
>
>
>
> +
>
> +  /**
>
> +    * Constructs an iterator where consecutive elements are accumulated as
>
>
> We should replace “iterator” with “collection”
> ------------------------------
>
> In src/main/scala/scala/collection/decorators/IterableDecorator.scala
> <https://github.com/scala/scala-collection-contrib/pull/4#discussion_r315087191>
> :
>
> > @@ -33,4 +33,46 @@ class IterableDecorator[C, I <: IsIterable[C]](coll: C)(implicit val it: I) {
>
>    def lazyFoldRight[B](z: B)(op: it.A => Either[B, B => B]): B =
>
>      it(coll).iterator.lazyFoldRight(z)(op)
>
>
>
> +
>
> +  /**
>
> +    * Constructs an iterator where consecutive elements are accumulated as
>
> +    * long as the output of f for each element doesn't change.
>
> +    * <pre>
>
> +    * Vector(1,2,2,3,3,3,2,2)
>
> +    * .splitBy(identity)
>
> +    * </pre>
>
> +    * produces
>
> +    * <pre>
>
> +    * Iterator(Vector(1),
>
>
> It should be Vector(Vector(1), instead of Iterator(Vector(1),
> ------------------------------
>
> In src/main/scala/scala/collection/decorators/IterableDecorator.scala
> <https://github.com/scala/scala-collection-contrib/pull/4#discussion_r315087445>
> :
>
> > +    * .splitBy(identity)
>
> +    * </pre>
>
> +    * produces
>
> +    * <pre>
>
> +    * Iterator(Vector(1),
>
> +    * Vector(2,2),
>
> +    * Vector(3,3,3),
>
> +    * Vector(2,2))
>
> +    * </pre>
>
> +    *
>
> +    * @param f the function to compute a key for an element
>
> +    * @tparam K the type of the computed key
>
> +    * @return an iterator of sequences of the consecutive elements with the
>
> +    *         same key in the original iterator
>
> +    */
>
> +  def splitBy[K, That, CC](f: it.A => K)(implicit bf: BuildFrom[C, it.A, That], bff: BuildFrom[C, That, CC]): CC = {
>
>
> What about naming the type parameters That1, That2? Or maybe C1 and C2?
> ------------------------------
>
> In src/main/scala/scala/collection/decorators/IterableDecorator.scala
> <https://github.com/scala/scala-collection-contrib/pull/4#discussion_r315088276>
> :
>
> > +    * <pre>
>
> +    * Vector(1,2,2,3,3,3,2,2)
>
> +    * .splitBy(identity)
>
> +    * </pre>
>
> +    * produces
>
> +    * <pre>
>
> +    * Iterator(Vector(1),
>
> +    * Vector(2,2),
>
> +    * Vector(3,3,3),
>
> +    * Vector(2,2))
>
> +    * </pre>
>
> +    *
>
> +    * @param f the function to compute a key for an element
>
> +    * @tparam K the type of the computed key
>
> +    * @return an iterator of sequences of the consecutive elements with the
>
> +    *         same key in the original iterator
>
>
> What about the following description?
>
> @return <https://github.com/return> a sequence of sequences of
> consecutive elements having the same key
>
> ------------------------------
>
> In src/main/scala/scala/collection/decorators/IterableDecorator.scala
> <https://github.com/scala/scala-collection-contrib/pull/4#discussion_r315088998>
> :
>
> > +    * @param f the function to compute a key for an element
>
> +    * @tparam K the type of the computed key
>
> +    * @return an iterator of sequences of the consecutive elements with the
>
> +    *         same key in the original iterator
>
> +    */
>
> +  def splitBy[K, That, CC](f: it.A => K)(implicit bf: BuildFrom[C, it.A, That], bff: BuildFrom[C, That, CC]): CC = {
>
> +    val builder = bff.newBuilder(coll)
>
> +
>
> +    val iterator = it(coll).iterator
>
> +    var init = bf.newBuilder(coll)
>
> +    if (iterator.hasNext) {
>
> +      var ref = iterator.next();
>
> +      init += ref
>
> +      while (iterator.hasNext) {
>
> +        val el = iterator.next();
>
> +        if (f(el) != f(ref)) {
>
>
> Should we put f(ref) in a val to compute it only once?
> ------------------------------
>
> In src/main/scala/scala/collection/decorators/IteratorDecorator.scala
> <https://github.com/scala/scala-collection-contrib/pull/4#discussion_r315089422>
> :
>
> > +
>
> +      override def next(): immutable.Seq[A] = {
>
> +        if (hasNext) {
>
> +          val seq = Vector.newBuilder[A]
>
> +          if (hdDefined) {
>
> +            seq += hd
>
> +          } else {
>
> +            hd = `this`.next()
>
> +            hdDefined = true
>
> +            seq += hd
>
> +          }
>
> +          var hadSameKey = true
>
> +          while (`this`.hasNext && hadSameKey) {
>
> +            val el = `this`.next()
>
> +            hdDefined = true
>
> +            if (f(el) == f(hd)) {
>
>
> Same comment here about f(hd)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala-collection-contrib/pull/4?email_source=notifications&email_token=AAAFTQ3OQOZAVD5OGKBL4DTQFJLDVA5CNFSM4FAUBXA2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCB4Y2MA#pullrequestreview-276401456>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAFTQ2YTQDMIAIESJ3OMBLQFJLDVANCNFSM4FAUBXAQ>
> .
>

jeantil commented:
I had some free time this morning thanks to kids sleeping in :)

julienrf commented:
Looks good to me, thanks! Do you mind squashing the commits into a single one?

jeantil commented:
No problem, I made separate commits to facilitate the review.

Le mar. 20 août 2019 à 10:17, Julien Richard-Foy <notifications@github.com>
a écrit :

> Looks good to me, thanks! Do you mind squashing the commits into a single
> one?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scala/scala-collection-contrib/pull/4?email_source=notifications&email_token=AAAFTQYEVH7SZK5QKKD3KQLQFOSBFA5CNFSM4FAUBXA2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4VO7HI#issuecomment-522907549>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAFTQ3PDOMBWWQT7VTUCGDQFOSBFANCNFSM4FAUBXAQ>
> .
>

jeantil commented:
Done I squashed everything to a single commit

julienrf commented:
:rocket:
